File: c/src/discount_applier.c
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include "discount_applier.h"



bool apply_v1(float discount, User* users, int userCount) {
    char message[MAX_MESSAGE_LENGTH];
    sprintf(message, "You've got a new discount of %.2f", discount);
    // Bug! Should be `(0, len(users))`
    for (int i = 1; i < userCount; i++) {
        User user = users[i];
        notify_discount(&user, message);
    }
    return true;
}

File: kotlin/src/main/kotlin/org/sammancoaching/DiscountApplier.kt
package org.sammancoaching

class DiscountApplier(private val notifier: Notifier) {
    fun applyV1(discount: Double, users: List<User>) {
        for (i in 1..<users.size) { // <- Bug, should be `i = 0`
            val message = java.lang.String.format("You've got a new discount of %f", discount)
            val user = users[i]
            notifier.notify(user, message)
        }
    }

    fun applyV2(discount: Double, users: List<User>) {
        for (i in users.indices) {
            val message = String.format("You've got a new discount of %f", discount)
            val user = users[0] // <- Bug, should be .get(i), not .get(0);
            notifier.notify(user, message)
        }
    }
}